Introduce an integration testing framework

ljedrz opened the pull request:
Builds on https://github.com/AleoHQ/snarkOS/pull/1285, including further changes outlined in https://github.com/AleoHQ/snarkOS/pull/1285#issuecomment-968844926.

This PR proposes a basic framework for integration tests (concentrating on the network first), including a thin/synthetic node implementation the full snarkOS node can be tested with/against.

Even though the synthetic node already works and is able to communicate with the full snarkOS node, there are still some wrinkles around the test and node setup that need to be ironed out, so not all of the tests are enabled yet. It would be best to agree on the overall design before proceeding.

niklaslong commented on testing/Cargo.toml:
```diff
@@ -0,0 +1,50 @@
+version = "0.1"
+
+[dependencies.tracing-subscriber]
+version = "0.2"
```
Why not `0.3` here?

ljedrz replied:
some copy-paste malfunction, it would seem :stuck_out_tongue:; will fix :+1:

niklaslong commented on testing/src/test_node.rs:
```diff
@@ -0,0 +1,310 @@
+    /// snarkOS recognizes nodes by their listening port; this collection facilitates
+    /// the peering experience during tests and aligns it with snarkOS logic.
+    // FIXME: use the `Disconnect` protocol to remove disconnected peers
+    peers: Arc<Mutex<HashSet<SocketAddr>>>,
```
Does this mean pea2pea's internal state is ignored in this case?

ljedrz replied:
no, it isn't - `pea2pea` just doesn't forbid "two-sided" connections, which are not needed here; all the state in this object is on top of the internal networking state of `pea2pea` (which will still forbid duplicate connections, obvious self-connections etc.)

niklaslong commented on testing/src/test_node.rs:
```diff
@@ -0,0 +1,310 @@
+        conn.writer().write_all(&msg).await?;
+
+        // Wait for the challenge response to come in.
+        conn.reader().read_exact(&mut buf[..4]).await?;
```
Might be good to have these lengths codefied as shared constants?

ljedrz replied:
fair; I was just drafting it quickly :+1:

niklaslong commented on testing/src/test_node.rs:
```diff
@@ -0,0 +1,310 @@
+    type Message = SnarkosMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        let mut buf = [0u8; 8 * 1024];
```
Do we have a max message size already?

ljedrz replied:
I haven't seen one - this is just an initial placeholder value, will add a `FIXME` :+1:

niklaslong commented on testing/src/test_node.rs:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
+        match message {
```
Suggestion: move the blocks into functions as this set is likely to grow?

ljedrz replied:
makes sense - will add a `TODO`

niklaslong commented on testing/tests/common/mod.rs:
```diff
@@ -0,0 +1,43 @@
+}
+
+#[macro_export]
+macro_rules! wait_until {
```
_Why hello there old friend, it's good to see you again._ 😏

niklaslong commented on testing/tests/network/manual_testing.rs:
```diff
@@ -0,0 +1,52 @@
+
+    // FIXME: snarkOS currently expects every node to be on this port, making it difficult
+    // to run local tests with multiple nodes.
+    const PORT: u16 = 4135;
```
Does snarkOS currently assign a port if it isn't specified?

ljedrz replied:
`snarkOS` currently uses and expects `4132` (for testnet2); this is just a placeholder that I used for manual tests (which required minor tweaks to the node) until these rules are relaxed (at least for test purposes)

niklaslong reviewed (approved):
LGTM, left some suggestions 👍

ljedrz commented:
great points, @niklaslong

ljedrz commented:
Addressed review comments.

ljedrz commented:
Rebased.

ljedrz commented:
The new tests now pass when run individually; I detected some issue with task accounting, so I temporarily disabled it.

howardwu commented:
Can we keep block locators in its original place? Alternatively, I can move it to snarkVM so it is out of the way here.

howardwu commented on src/node.rs:
```diff
@@ -35,12 +41,21 @@ pub struct Node {
     pub network: u16,
+    /// The listener IP of the node.
+    #[structopt(default_value = "0:0:0:0", long = "ip")]
+    pub ip: String,
```
Why is this necessary?

ljedrz replied:
Two reasons: 
- users utilizing multiple networks might not want the node to be reachable on all of them
- we don't want test nodes to connect to the testnet, and this allows them to be `127.0.0.1`

niklaslong replied:
Oh, I see this is a `String` (ignore my earlier `Copy` comments), why not use `IpAddr`?

howardwu commented on src/node.rs:
```diff
@@ -35,12 +41,21 @@ pub struct Node {
-    #[structopt(long = "node")]
-    pub node: Option<u16>,
+    #[structopt(long = "listener-port")]
+    pub listener_port: Option<u16>,
```
Please change back to `node`

howardwu commented on src/node.rs:
```diff
@@ -35,12 +41,21 @@ pub struct Node {
+    pub listener_port: Option<u16>,
+    /// Disable the RPC server.
+    #[structopt(long = "disable-rpc")]
+    pub disable_rpc: bool,
```
This is very messy, let's take another design look.

ljedrz replied:
What is messy about it? Not spawning an RPC server in non-RPC tests makes them a lot faster, so it's a desirable feature. Alternatively I could disable that block directly using `#[cfg(not(feature = "test"))]`.

howardwu commented on src/node.rs:
```diff
@@ -35,12 +41,21 @@ pub struct Node {
+    pub disable_rpc: bool,
+    /// The IP address of the RPC server.
+    #[structopt(default_value = "0:0:0:0", long = "rpc-ip")]
+    pub rpc_ip: String,
```
Why is this necessary?

ljedrz replied:
See https://github.com/AleoHQ/snarkOS/pull/1286#discussion_r750972245.

niklaslong replied:
This should also be an `IpAddr`?

ljedrz replied:
I didn't think `structopt` was able to interpret the default value "on the fly" like that, but it appears that it is :+1:.

howardwu commented on src/node.rs:
```diff
@@ -35,12 +41,21 @@ pub struct Node {
-    #[structopt(long = "rpc")]
-    pub rpc: Option<u16>,
+    #[structopt(long = "rpc-port")]
+    pub rpc_port: Option<u16>,
```
Please change back to `rpc`

howardwu commented:
I would prefer `Environment` remain in a primary crate here. It is not "external" to anything here

ljedrz commented:
> I would prefer Environment remain in a primary crate here. It is not "external" to anything here

I wanted to propose this "split" early on (as it might become more difficult with time), but in general, it's a matter of being able to build outside node (e.g. a dedicated crawler) and RPC implementations faster, without having to build entire snarkOS. The term "external" refers to the fact that these are (except for RPC, which I left for later) objects facing "the external world", i.e. exposed to the network.

I can revert these changes for now, but this is a feature that had been proposed before and which provides clear benefits for the ecosystem, so I would recommend considering it for the future.

ljedrz committed:
feat: introduce a network integration test framework

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz commented:
Rebased and addressed most comments.

note: I checked the alternative solution to disabling the RPC server mentioned in https://github.com/AleoHQ/snarkOS/pull/1286#discussion_r750987265, but it was a lot messier - it required multiple imports and `const`s to obtain the extra attribute.

ljedrz committed:
tests: add connection cleanup test drafts

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz committed:
chore: adjust Cargo.lock

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz committed:
tests: add additional network integration tests

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz committed:
tests: add a connection limit test

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz commented:
Rebased and ready for re-review.

niklaslong commented on src/network/server.rs:
```diff
@@ -54,29 +56,35 @@ impl<N: Network, E: Environment> Server<N, E> {
-
-        // Initialize the tasks handler.
-        let mut tasks = Tasks::new();
+        #[cfg(not(feature = "test"))]
```
Might be good to comment on these few lines a little?

niklaslong commented on src/network/server.rs:
```diff
@@ -54,29 +56,35 @@ impl<N: Network, E: Environment> Server<N, E> {
-        miner: Option<Address<N>>,
-    ) -> Result<Self> {
+    pub(crate) async fn initialize(node: &Node, miner: Option<Address<N>>, mut tasks: Tasks<task::JoinHandle<()>>) -> Result<Self> {
+        let node_ip = node.ip.clone();
```
I'm assuming this is `Copy`?

niklaslong commented on src/network/server.rs:
```diff
@@ -54,29 +56,35 @@ impl<N: Network, E: Environment> Server<N, E> {
+    pub(crate) async fn initialize(node: &Node, miner: Option<Address<N>>, mut tasks: Tasks<task::JoinHandle<()>>) -> Result<Self> {
+        let node_ip = node.ip.clone();
+        let node_port = node.node.unwrap_or(E::DEFAULT_NODE_PORT);
+        let rpc_ip = node.rpc_ip.clone();
```
ditto ☝️

niklaslong commented on src/node.rs:
```diff
@@ -105,25 +119,32 @@ impl Node {
+        }
+
+        let server_task = task::spawn(async move {
+            let _server = server;
```
I'm guessing this is to avoid prematurely dropping the `server`?

niklaslong commented on testing/src/snarkos_node.rs:
```diff
@@ -0,0 +1,73 @@
+        // Procure a free port number for the snarkOS node.
+        // FIXME: due to there being a delay between the port's discovery and its binding by the node, this
+        // method can cause an `AddrInUse` error to occur when multiple tests are run at the same time; only
+        // introspection of a ready node can fully avoid this.
```
I suppose we can run them with `--test-threads 1` in the meantime?

ljedrz replied:
Some of those tests require the executor to be multi-threaded, though; these failures are quite rare.

niklaslong commented on testing/src/snarkos_node.rs:
```diff
@@ -0,0 +1,73 @@
+        SnarkosNode::with_args(&["--node", &free_port.to_string()]).await
+    }
+
+    pub async fn with_args(extra_args: &[&str]) -> Self {
```
Some doc here might be good (let's start early 😉), especially as some args are permanent?

niklaslong commented on testing/tests/network/basic_connectivity.rs:
```diff
@@ -0,0 +1,137 @@
+    let test_node1 = TestNode::default().await;
+
+    // Ensure that the nodes have no active connections.
+    crate::wait_until!(1, test_node0.node().num_connected() == 0 && test_node1.node().num_connected() == 0);
```
```suggestion
    wait_until!(1, test_node0.node().num_connected() == 0 && test_node1.node().num_connected() == 0);
```
It's imported up top ☝️

niklaslong commented on testing/tests/network/basic_connectivity.rs:
```diff
@@ -0,0 +1,137 @@
+    SnarkosNode::with_args(&["--node", "0", "--connect", &test_node_addr.to_string()]).await;
+
+    // The snarkOS node should have connected to the test node.
+    crate::wait_until!(5, test_node.node().num_connected() != 0);
```
ditto

niklaslong commented on testing/tests/network/basic_connectivity.rs:
```diff
@@ -0,0 +1,137 @@
+async fn node_cant_connect_to_itself() {}
+
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "TODO: currently not possible to connect on demand"]
```
Nice 👍

ljedrz committed:
tweak: apply review suggestions

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz committed:
tweak: improve address creation

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz commented:
Applied review suggestions by @niklaslong.

ljedrz committed:
fix: add a pending future in main

Signed-off-by: ljedrz <ljedrz@gmail.com>

ljedrz commented:
Superseded by https://github.com/AleoHQ/snarkOS/pull/1311.
